### PR TITLE
Add new reduction mode  in kl_div

### LINKED ISSUE #6622: Possible KL_loss bug on output dimension average.
HI I find the KL_loss compute by pytorch is always small than tensorflow. Then I check the defination of KL loss https://en.wikipedia.org/wiki/Kullback%E2%80%93Leibler_divergence, and also the decuments of the http://pytorch.org/docs/master/nn.html, It shows **By default, the losses are averaged for each minibatch over observations as well as over dimensions. However, if the field size_average is set to False, the losses are instead summed.** in pytorch. The problem is the KL loss from wikipedia is not using average dimensions. Any misunderstanding for KL loss implement by pytorch.

### PR DESCRIPTION
**Pull request by ailzhang**
Fixes #6622 . 
We used to average over all elements for kl divergence, which is not aligned with its math definition. 
This PR corrects the default reduction behavior of KL divergence that it now naverages over batch dimension. 

- In KL, default behavior `reduction=mean` averages over batch dimension. While for most other loss functions, `reduction=mean` averages over all elements. 
- We used to support scalar tensor as well. For BC purpose, we still support it, no reduction is performed on scalar tensor. 
- Added a new reduction mode called `batchmean` which has the correct behavior for KL. Add a warning to make `batchmean` as default for KL instead of `mean` in next major release. 
- [deprecated]I chose to not add a new reduction option, since "mean over batch dimension" is kinda special, and it only makes sense in few cases like KL. We don't want to explain why there's a option "batchmean" but it's not applicable for all other functions. I'm open to discussion on this one, as I cannot think of a perfect solution for this.

### CODE REVIEW

**Review by ssnl — COMMENTED** [review]
can we get an expect test that this averages along batch dimension?

**Comment by ailzhang** [comment]
@SsnL  I added a test to compare it with `reduction='none'`. Let me know what you think, thanks!

File: torch/nn/functional.py [hunk 1]
```diff
@@ -1719,15 +1719,23 @@ def kl_div(input, target, size_average=None, reduce=None, reduction='mean'):
...
             batch element instead and ignores :attr:`size_average`. Default: ``True``
         reduction (string, optional): Specifies the reduction to apply to the output:
-            'none' | 'mean' | 'sum'. 'none': no reduction will be applied,
-            'mean': the sum of the output will be divided by the number of
-            elements in the output, 'sum': the output will be summed. Note: :attr:`size_average`
-            and :attr:`reduce` are in the process of being deprecated, and in the meantime,
-            specifying either of those two args will override :attr:`reduction`. Default: 'mean'
+            'none' | 'batchmean' | 'sum' | 'mean'. 'none': no reduction will be applied,
+            'batchmean': the sum of the output will be divided by the number of
+            batches in the output, 'sum': the output will be summed, 'mean': the output will be
```
Comment (soumith) [inline torch/nn/functional.py]: "the number of batches" -> "the batch size" or "the number of elements in the input batch"

File: aten/src/ATen/core/Reduction.h [hunk 2]
```diff
@@ -9,6 +9,7 @@ enum Reduction {
   None,             // Do not reduce
   Mean,             // (Possibly weighted) mean of losses
   Sum,              // Sum losses
+  BatchMean,        // Mean over batches. = Sum / batchsize
```
Comment (SsnL) [inline aten/src/ATen/core/Reduction.h]: Can't you just hack this up in python? It should only be a workaround for kl_div and doesn't make sense for other losses.

**Review by ssnl — CHANGES_REQUESTED** [review]
I don't think `batch_mean` should be an enum element. It will also introduce weird behavior for the losses still defined in TH if used with `batch_mean` since if-statements there are not written with `batch_mean` in mind.

**Review by ssnl — APPROVED** [review]
I'm now wondering if reduction='none' is wrong.. But this generally LGTM.
